res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+            this.forEachComponentContainer(function(container) {
+                container.unshift();
+            }, false);
+
+            // Unset isShifted for each ComponentContainer recursively
+            this.isShifted = false;
+        }
+        shutdown() {
+            this.forEachComponent(function(component) {
+                if (component.shutdown !== undefined
+                    && typeof component.shutdown === "function") {
+                    component.shutdown();
+                }
+            });
+        }
+    };
+
+    const Deck = class extends ComponentContainer {
+        constructor(deckNumbers) {
+            super();
+            if (deckNumbers !== undefined) {
+                if (Array.isArray(deckNumbers)) {
+                    // These must be unique to each instance,
+                    // so they cannot be in the prototype.
+                    this.currentDeck = "[Channel" + deckNumbers[0] + "]";
+                    this.deckNumbers = deckNumbers;
+                } else if (typeof deckNumbers === "number" &&
+                        Math.floor(deckNumbers) === deckNumbers &&
+                        isFinite(deckNumbers)) {
+                    this.currentDeck = "[Channel" + deckNumbers + "]";
+                    this.deckNumbers = [deckNumbers];
+                }
+            } else {
+                throw "new Deck() called without specifying any deck numbers";
+            }
+        }
+        setCurrentDeck(newGroup) {
+            this.currentDeck = newGroup;
+            this.reconnectComponents(function(component) {
+                if (component.group === undefined
+                      || component.group.search(script.channelRegEx) !== -1) {
+                    component.group = newGroup;
+                } else if (component.group.search(script.eqRegEx) !== -1) {
+                    component.group = "[EqualizerRack1_" + newGroup + "_Effect1]";
+                } else if (component.group.search(script.quickEffectRegEx) !== -1) {
+                    component.group = "[QuickEffectRack1_" + newGroup + "]";
+                }
+                // Do not alter the Component's group if it does not match any of those RegExs.
+
+                if (component instanceof EffectAssignmentButton) {
+                    // The ControlObjects for assigning decks to effect units
+                    // indicate the effect unit with the group and the deck with the key,
+                    // so change the key here instead of the group.
+                    component.inKey = "group_" + newGroup + "_enable";
+                    component.outKey = "group_" + newGroup + "_enable";
+                }
+            });
+        }
+        toggle() {
+            // cycle through deckNumbers array
+            let index = this.deckNumbers.indexOf(parseInt(
+                script.channelRegEx.exec(this.currentDeck)[1]
+            ));
+            if (index === (this.deckNumbers.length - 1)) {
```

Comment:
simpler:  index = (index + 1) % (this.deckNumbers.length - 1)

Reply:
usually, `%` is expensive, but this code is not hot, so sure.

Reply:
The - 1 doesn't make sense in this context anymore. The version I pushed should be correct.